plugin-server/src/main/ingestion-queues/session-recording-v2/retention/retention-aware-batch-writer.ts

```diff
@@ -0,0 +1,93 @@
+import { RetentionPeriod } from '../types'
+import { RetentionService } from './retention-service'
+
+class RetentionAwareBatchFileWriter implements SessionBatchFileWriter {
+    private storageMap: { [key in RetentionPeriod]: SessionBatchFileStorage }
+    private writerMap: { [key in RetentionPeriod]: SessionBatchFileWriter | null }
+
+    constructor(
+        private readonly s3: S3Client,
+        private readonly bucket: string,
+        private readonly prefix: string,
+        private readonly timeout: number,
+        private readonly retentionService: RetentionService
+    ) {
+        this.storageMap = ValidRetentionPeriods.reduce(
+            (storage, retentionPeriod) => {
+                storage[retentionPeriod] = new S3SessionBatchFileStorage(
+                    this.s3,
+                    this.bucket,
+                    retentionPeriod === 'legacy' ? 'session_recording_batches' : `${this.prefix}/${retentionPeriod}`,
+                    this.timeout
+                )
+                return storage
+            },
+            {} as { [key in RetentionPeriod]: SessionBatchFileStorage }
+        )
+
+        this.writerMap = ValidRetentionPeriods.reduce(
+            (writers, retentionPeriod) => {
+                writers[retentionPeriod] = null
+                return writers
+            },
+            {} as { [key in RetentionPeriod]: SessionBatchFileWriter | null }
+        )
+    }
+
+    public async writeSession(sessionData: SessionData): Promise<WriteSessionResult> {
+        const retentionPeriod = await this.retentionService.getSessionRetention(
+            sessionData.teamId,
+            sessionData.sessionId
+        )
+
+        let writer = this.writerMap[retentionPeriod]
+
+        if (writer === null) {
+            const storage = this.storageMap[retentionPeriod]
+            writer = storage.newBatch()
+            this.writerMap[retentionPeriod] = writer
+        }
+
+        return writer.writeSession(sessionData)
+    }
+
+    public async finish(): Promise<void> {
+        await Promise.all(
+            ValidRetentionPeriods.map(async (retentionPeriod) => {
+                const writer = this.writerMap[retentionPeriod]
+
+                if (writer !== null) {
+                    await writer.finish()
+                }
+            })
+        )
+    }
+}
+
+export class RetentionAwareStorage implements SessionBatchFileStorage {
+    constructor(
+        private readonly s3: S3Client,
+        private readonly bucket: string,
+        private readonly prefix: string,
+        private readonly timeout: number = 5000,
+        private readonly retentionService: RetentionService
+    ) {}
+
+    public newBatch(): RetentionAwareBatchFileWriter {
+        return new RetentionAwareBatchFileWriter(this.s3, this.bucket, this.prefix, this.timeout, this.retentionService)
+    }
+
+    public checkHealth(): Promise<boolean> {
+        return Promise.resolve(true)
```

Comment:
suggestion: If we can, it would be great to have this healthcheck. As we're using the same bucket, just different prefixes, we could just pick the first S3 writer to run the healthcheck.

The healthchecks resolve some of the issues with minio on local dev, where uploads don't work before sending a healthcheck.